### ISSUE TITLE
Raw strings inferred as literal types instead of strings, which cause spurious type errors.

### ISSUE THREAD

**Issue by FedericoV** [microsoft/pyright#274]
**Describe the bug**

Under some conditions, pyright thinks that raw strings are literal types, and those cannot be used in place of strings in code.

**To Reproduce**
Steps to reproduce the behavior.

```
    trash_files = [
        os.path.join(os.path.dirname(response_keys[0]), "eval2.csv.out")]
```
gives this error:
```
No overloads for 'os.path.join(os.path.dirname(response_keys[0]), "eval2.csv.out")' match parameters
  Argument types: (Union[str, bytes], Literal['eval2.csv.out'])
```

**Expected behavior**

It should recognize that a literal string type can be used in place of a string.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```
def foo(self) -> str:
    return 3
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

Name: Pyright
Id: ms-pyright.pyright
Description: VS Code static type checking for Python
Version: 1.0.67
Publisher: ms-pyright

**Additional context**

**Comment by erictraut** [comment]
Here's what's happening. It has nothing to do with literal types.

I'm guessing that `response_keys` is is of an unknown type — probably because it is not annotated. The `os.path.dirname` method is annotated as follows:
```
def dirname(path: AnyStr) -> AnyStr: ...
```
where AnyStr is defined as:
```
AnyStr = TypeVar('AnyStr', str, bytes)
```
You're passing a variable of unknown type to the dirname method. The most that pyright can assume is that it must be either a str or a bytes, and that the result must be `Union[str, bytes]`. 

Now you're passing a value of type `Union[str, bytes]` as the first argument to `os.path.join`. This method is annotated as follows:
```
@overload
def join(path: _StrPath, *paths: _StrPath) -> Text: ...

@overload
def join(path: _BytesPath, *paths: _BytesPath) -> bytes: ...
```
This means the first parameter must be either a str or a bytes to match one of these overloads. It can't be both (a union).

Interestingly enough, the annotated definition of `join` appears to have changed as of python 3.6. Prior to that, it was defined as:
```
def join(path: AnyStr, *paths: AnyStr) -> AnyStr: ...
```
And that allows the union. I'm not sure why the change was made for python 3.6 and newer.

The workaround is to ensure that `response_keys` has a defined type of either `Iterator[str]` or `Iterator[bytes]`.
